File: android/a-kart/app/src/main/java/com/frogdesign/akart/util/BmpToYUVToARToolkitConverter2.java
package com.frogdesign.akart.util;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.ImageFormat;
import android.renderscript.Allocation;
import android.renderscript.Element;
import android.renderscript.RenderScript;
import android.renderscript.Type;

import com.frogdesign.akart.ScriptC_argb_to_yuv;

import org.artoolkit.ar.base.ARToolKit;
import org.artoolkit.ar.base.Utils;

import rx.functions.Func1;


public class BmpToYUVToARToolkitConverter2 implements Func1<Bitmap, Boolean> {
    private byte[] yuvBuffer = null;

    private Allocation inAllocation;
    private Allocation outAllocation;
    private RenderScript rs;
    private ScriptC_argb_to_yuv script;

    public BmpToYUVToARToolkitConverter2(Context ctx) {
        rs = RenderScript.create(ctx);
        script = new ScriptC_argb_to_yuv(rs);
    }

    private void checkForBuffers(Bitmap b) {

        int w = b.getWidth();
        int h = b.getHeight();
        int yuvLength = Utils.yuvByteLength(w, h);
        if (yuvBuffer == null || yuvBuffer.length != yuvLength) {

            inAllocation = Allocation.createFromBitmap(rs, b,
                    Allocation.MipmapControl.MIPMAP_NONE, Allocation.USAGE_SCRIPT);
            script.set_gInImage(inAllocation);

            yuvBuffer = new byte[yuvLength];
            Type.Builder tbIn = new Type.Builder(rs, Element.U8(rs));
            tbIn.setX(w);
            tbIn.setY(h);
            tbIn.setYuvFormat(ImageFormat.NV21);

            outAllocation = Allocation.createSized(rs, Element.U8(rs), yuvBuffer.length, Allocation.USAGE_SCRIPT);

            script.bind_outBytes(outAllocation);
            script.set_width(w);
            script.set_height(h);
            script.set_frameSize(w * h);
        }
    }

    @Override
    public Boolean call(Bitmap inBitmap) {
        if (inBitmap == null) return false;
        //Log.i(PlayActivity.TAG, "convert" + isMainThread())
        checkForBuffers(inBitmap);

        inAllocation.copyFrom(inBitmap);

        script.invoke_filter();

        outAllocation.copyTo(yuvBuffer);

        if (ARToolKit.getInstance().nativeInitialised()) {
            ARToolKit.getInstance().convertAndDetect(yuvBuffer);
        }
        return true;
    }
}

File: android/a-kart/app/src/main/kotlin/com/frogdesign/akart/util/BmpToYUVToARToolkitConverter2.kt
package com.frogdesign.akart.util

import android.content.Context
import android.graphics.Bitmap
import android.graphics.ImageFormat
import android.renderscript.Allocation
import android.renderscript.Element
import android.renderscript.RenderScript
import android.renderscript.Type
import android.util.Log

import com.frogdesign.akart.ScriptC_argb_to_yuv

import org.artoolkit.ar.base.ARToolKit

import rx.functions.Func1


class BmpToYUVToARToolkitConverter2(ctx: Context) : Func1<Bitmap, Boolean> {
    private var yuvBuffer: ByteArray? = null

    private var inAllocation: Allocation? = null
    private var outAllocation: Allocation? = null
    private val rs: RenderScript
    private val script: ScriptC_argb_to_yuv

    init {
        rs = RenderScript.create(ctx)
        script = ScriptC_argb_to_yuv(rs)
    }

    private fun checkForBuffers(b: Bitmap) {

        val w = b.width
        val h = b.height
        val yuvLength = yuvByteLength(w, h)
        if (yuvBuffer == null || yuvBuffer!!.size != yuvLength) {

            inAllocation = Allocation.createFromBitmap(rs, b,
                    Allocation.MipmapControl.MIPMAP_NONE, Allocation.USAGE_SCRIPT)
            script._gInImage = inAllocation

            yuvBuffer = ByteArray(yuvLength)

            outAllocation = Allocation.createSized(rs, Element.U8(rs), yuvBuffer!!.size, Allocation.USAGE_SCRIPT)

            script.bind_outBytes(outAllocation)
            script._width = w
            script._height = h
            script._frameSize = w * h
        }
    }

    override fun call(inBitmap: Bitmap?): Boolean? {
        if (inBitmap == null) return false
        //Log.i(PlayActivity.TAG, "convert" + isMainThread())
        checkForBuffers(inBitmap)

        inAllocation!!.copyFrom(inBitmap)

        //long startT = System.nanoTime();
        script.invoke_filter()
        //script.forEach_rgbToYuv(inAllocation);
        //long dt = System.nanoTime() - startT;
        //Log.i("TIME", Long.toString(dt));

        outAllocation!!.copyTo(yuvBuffer)

        if (ARToolKit.getInstance().nativeInitialised()) {
            ARToolKit.getInstance().convertAndDetect(yuvBuffer)
        }
        return true
    }
}